### PR TITLE
Filter Emission Factor Value

### PR DESCRIPTION
**Pull request by isaaccodekill**
- Convert all units to default-units in calculations 
- Display Emission factor units based on selected extra field values

### CODE REVIEW

File: app/src/app/api/v0/emissions-factor/route.ts [hunk 1]
```diff
@@ -101,6 +92,17 @@ export const POST = apiHandler(async (req: NextRequest, _context: {}) => {
...
+  if (methodologyId) {
+    if (methodologyId.includes("fuel-combustion"))
+      whereClause.methodologyName = "fuel-combustion-consumption";
+    if (methodologyId.includes("scaled"))
+      whereClause.methodologyName = "sampling-scaled-data";
+    if (methodologyId.includes("modeled-data"))
+      whereClause.methodologyName = "modeled-data";
+  }
+
+  console.log(whereClause, "where");
```
Comment (lemilonkh) [inline app/src/app/api/v0/emissions-factor/route.ts]: Remove logging

File: app/src/backend/UnitConversionService.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
+// Purpose: Service to convert units between different systems.
+
+import createHttpError from "http-errors";
+
+export default class UnitConversionService {
+  private static conversionTable: { [key: string]: { [key: string]: number } } =
+    {
+      "units-cubic-meters": {
+        "units-gallons": 0.00378541, // 1 gallon = 0.00378541 cubic meters
```
Comment (lemilonkh) [inline app/src/backend/UnitConversionService.ts:9]: Maybe we should use decimal.js for conversions here at some point, so we don't run into floating point inaccuracy issues... This many decimals are probably handle-able by JS Numbers, but it would be nice to check if the resulting precision is sufficient in a later QA step.

File: app/src/backend/formulas.ts [hunk 3]
```diff
@@ -115,6 +117,80 @@ const DEFAULT_INCOME_GROUP_FRACTIONS: Record<string, number> = {
...
+  if (activity?.["default-units"]) {
+    const val = data[activity?.["activity-title"] as string];
+    const fuelTypeKey = Object.keys(data).find((key) =>
+      key.includes("fuel-type"),
+    );
+    const fuelType = data[fuelTypeKey as string];
+    data[activity?.["activity-title"] as string] = new Decimal(
+      UnitConversionService.convertUnits(
+        val,
+        data[`${activity?.["activity-title"]}-unit`],
```
Comment (lemilonkh) [inline app/src/backend/formulas.ts:168]: Could you maybe extra this key into a variable beforehand instead of doing string interpolation in the object access in the function call?

File: app/src/backend/formulas.ts [hunk 4]
```diff
@@ -321,8 +408,14 @@ export function handleVkt1Formula(
 
 export function handleMethaneCommitmentFormula(
   activityValue: ActivityValue,
+  inventoryValue: InventoryValue,
 ): Gas[] {
-  const data = activityValue.activityData;
+  const data = convertDataToDefaultUnit(
+    // use convert all the values
+    activityValue,
+    inventoryValue.inputMethodology as string,
```
Comment (lemilonkh) [inline app/src/backend/formulas.ts:416]: Instead of converting the type it might be good to handle null cases in some place beforehand?

File: app/src/backend/formulas.ts [hunk 5]
```diff
@@ -466,11 +571,18 @@ export function handleIndustrialWasteWaterFormula(
   return [{ gas: "CH4", amount }];
 }
 
+// TODO how correct is this formula ?
```
Comment (lemilonkh) [inline app/src/backend/formulas.ts:574]: Yes I think Mau's calculations confirmed we have some work to do on it. We will have tickets for the upcoming sprint for it.

File: app/src/components/Modals/activity-modal/activity-form-modal.tsx [hunk 6]
```diff
@@ -85,6 +84,15 @@ const AddActivityModal: FC<AddActivityModalProps> = ({
     methodology: methodology,
   });
 
+  const { emissionsFactorTypes, emissionsFactorsLoading } = useEmissionFactors({
```
Comment (lemilonkh) [inline app/src/components/Modals/activity-modal/activity-form-modal.tsx:87]: Calling it `areEmissionsFactorsLoading` might make it more clear that it's a boolean.

File: app/src/components/Modals/activity-modal/activity-form-modal.tsx [hunk 7]
```diff
@@ -348,6 +306,7 @@ const AddActivityModal: FC<AddActivityModalProps> = ({
           <ModalCloseButton marginTop="10px" />
           <ActivityModalBody
             emissionsFactorTypes={emissionsFactorTypes}
+            loadingEmissionFactors={emissionsFactorsLoading}
```
Comment (lemilonkh) [inline app/src/components/Modals/activity-modal/activity-form-modal.tsx:309]: Two different names for the same field are a bit confusing

File: app/src/components/Modals/activity-modal/activity-modal-body.tsx [hunk 8]
```diff
@@ -113,12 +119,18 @@ const ActivityModalBody = ({
...
 
   const unitValue = useWatch({
     control,
-    name: `activity.${title}Unit` as any,
+    name: `activity.${title}-unit` as any,
+  });
+
+  const emissionFactorTypeValue = useWatch({
+    control,
+    name: `activity.emissionFactorType` as any,
```
Comment (lemilonkh) [inline app/src/components/Modals/activity-modal/activity-modal-body.tsx:133]: As any should not be required here? Same as back tick string.

File: app/src/components/formatted-number-input.tsx [hunk 9]
```diff
@@ -57,6 +59,12 @@ function FormattedNumberInput({
   return (
     <Controller
       control={control}
+      rules={{
+        required: t("value-required"),
+        validate: (value) => {
+          return (value !== "" && !isNaN(value)) || t("value-required");
```
Comment (lemilonkh) [inline app/src/components/formatted-number-input.tsx:65]: Remove debug logging

File: app/src/hooks/activity-value-form/use-emission-factors.ts [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+    ...source,
+    gasValuesByGas: Object.keys(source.gasValuesByGas).reduce(
+      (acc, currentValue) => {
+        return {
+          ...acc,
+          [currentValue]: {
+            gasValues: uniqBy(
+              source.gasValuesByGas[currentValue].gasValues,
+              "emissionsPerActivity",
+            ).filter((factor) => ["kg/m3", "kg/kWh"].includes(factor.units)), // filter only emissions that have kg/m3 or kg/kWh as the unit
```
Comment (lemilonkh) [inline app/src/hooks/activity-value-form/use-emission-factors.ts:67]: Why only these?

File: app/src/hooks/activity-value-form/use-emission-factors.ts [hunk 11]
```diff
@@ -0,0 +1,155 @@
+// take the methodology selected,
```
Comment (lemilonkh) [inline app/src/hooks/activity-value-form/use-emission-factors.ts:1]: What is this comment for?

File: app/src/hooks/activity-value-form/use-emission-factors.ts [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+      return [];
+    }
+
+    // now that we have three or more emission factors, we want to reduce it down to a collection of gases per dataset
+    return reduceEmissionsToUniqueSourcesAndUnits(emissionsFactors);
+  }, [emissionsFactors]);
+
+  // when the data changes we wanna set the emissions factor types to the first selection that exists.
+  useEffect(() => {
+    console.log(emissionsFactorTypes, "they changed");
```
Comment (lemilonkh) [inline app/src/hooks/activity-value-form/use-emission-factors.ts:143]: Remove debug logging

File: app/src/hooks/activity-value-form/use-emission-factors.ts [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+    // if (emissionsFactorTypes.length > 0) {
+    //   setValue("activity.emissionFactorType", emissionsFactorTypes[0].id);
+    // }
```
Comment (lemilonkh) [inline app/src/hooks/activity-value-form/use-emission-factors.ts:146]: Does this need to be re-enabled?

File: app/src/services/api.ts [hunk 14]
```diff
@@ -23,636 +23,637 @@ import type {
   UserInfoResponse,
   UserInviteResponse,
 } from "@/util/types";
-import type { GeoJSON } from "geojson";
-import { createApi, fetchBaseQuery } from "@reduxjs/toolkit/query/react";
+import type {GeoJSON} from "geojson";
```
Comment (lemilonkh) [inline app/src/services/api.ts:26]: Can you run Prettier on this file to remove the unnecessary formatting changes?

**Review by lemilonkh — CHANGES_REQUESTED** [review]

File: app/src/backend/UnitConversionService.ts (hunk 2 above)
Comment (isaaccodekill) [reply app/src/backend/UnitConversionService.ts:9]: Noted

File: app/src/hooks/activity-value-form/use-emission-factors.ts (hunk 13 above)
Comment (isaaccodekill) [reply app/src/hooks/activity-value-form/use-emission-factors.ts:146]: Not really

File: app/src/hooks/activity-value-form/use-emission-factors.ts (hunk 10 above)
Comment (isaaccodekill) [reply app/src/hooks/activity-value-form/use-emission-factors.ts:67]: These are the units that every emission factor should be in. I and @mfonsecaOEF  agreed that the emissions factors should all be in the same units. 

@mfonsecaOEF  is still working on converting all EFs to these units. But until she's done,  i'm filtering out EFs that are not these units so that the calculations will be accurate

File: app/src/backend/formulas.ts (hunk 4 above)
Comment (isaaccodekill) [reply app/src/backend/formulas.ts:416]: the input methodology will always be required for conversions because we need to access the methodology in the manual input json file

Comment (lemilonkh) [reply app/src/backend/formulas.ts:416]: Yes, I meant doing a null check and throwing/ logging an error here instead of effectively using the `!` Typescript postfix operator by using `as string`

Comment (lemilonkh) [reply app/src/backend/formulas.ts:416]: Checking it and returning/ throwing in that case means it is no longer considered nullable in the current block by the TS compiler, meaning you can leave out `as string` in all of those places.

**Review by lemilonkh — APPROVED** [review]
Thanks ✨

File: app/src/backend/formulas.ts (hunk 4 above)
Comment (isaaccodekill) [reply app/src/backend/formulas.ts:416]: noted and corrected. THanks
